### PR TITLE
feat: [PL-26945]: Added migration task for service account token

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
Migration of the field encryptedServiceAccountToken for the customer Kroger(accId- yTbabkw4SdCPZdnEWynUNg) for erroneous settingAttributes. Jira- https://harness.atlassian.net/browse/PL-26945
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 2 Issues 
* Code issue 1 Code issues 
* Anti-pattern 1 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37231)

**Comment by tiwaritejas** [comment]
trigger messagecheck

**Comment by mohitg0795** [comment]
*[`390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java` line 70 at r1](https://reviewable.io/reviews/harness/harness-core/37231#-NBMot5RCJ7u-7R_Xbyy:-NBMot5T86TZXN4YRHZB:b-52lye8) ([raw file](https://github.com/harness/harness-core/blob/fa49258071b43d49f1de1c0c9f58acbe2df94655/390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java#L70)):*
> ```Java
>         }
>       }
>     }
> ```

lets add catch block and log exception if anything goes wrong.

**Review by gn-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 3 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java` line 64 at r1](https://reviewable.io/reviews/harness/harness-core/37231#-NBPRcD8Dhct_SR071Co:-NBPRcD8Dhct_SR071Cp:b-wa4bz0) ([raw file](https://github.com/harness/harness-core/blob/fa49258071b43d49f1de1c0c9f58acbe2df94655/390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java#L64)):*
> ```Java
>         UpdateOperations<SettingAttribute> ops = mongoPersistence.createUpdateOperations(SettingAttribute.class);
>         String currentToken =
>             tokenIdsMap.get(((KubernetesClusterConfig) settingAttribute.getValue()).getEncryptedServiceAccountToken());
> ```

Something looks fishy - you are checking presence of "(KubernetesClusterConfig) settingAttribute.getValue()).getEncryptedServiceAccountToken()" inside tokenIds and then looking up the same in tokenIdsMap -- is that expected?

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 3 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java` line 64 at r1](https://reviewable.io/reviews/harness/harness-core/37231#-NBPRcD8Dhct_SR071Co:-NBQac4d-bRJ-yrjaecg:bq1vrqw) ([raw file](https://github.com/harness/harness-core/blob/fa49258071b43d49f1de1c0c9f58acbe2df94655/390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java#L64)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

Something looks fishy - you are checking presence of "(KubernetesClusterConfig) settingAttribute.getValue()).getEncryptedServiceAccountToken()" inside tokenIds and then looking up the same in tokenIdsMap -- is that expected?
</blockquote></details>

Yes, this is expected. The customer has incorrectly set the encryptedServiceToken field as secret name instead of secret id. Now, the tokenIdsMap maintains a map of secret name to secret ids. So, for each record, I check if the encryptedServiceToken indeed holds a valid secret id, and if so, I continue. Else, I go ahead and replace the secret name by secret id by fetching it from the tokenIdsMap.
___
*[`390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java` line 70 at r1](https://reviewable.io/reviews/harness/harness-core/37231#-NBMot5RCJ7u-7R_Xbyy:-NBQdIgTCa3m4yUBkQ4V:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fa49258071b43d49f1de1c0c9f58acbe2df94655/390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java#L70)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

lets add catch block and log exception if anything goes wrong.
</blockquote></details>

Done.

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 3 file(s) to review. 2 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java` line 34 at r2](https://reviewable.io/reviews/harness/harness-core/37231#-NBRl17n1n_dzOcf8DWD:-NBRl17n1n_dzOcf8DWE:b2dn17e) ([raw file](https://github.com/harness/harness-core/blob/efbb95c7e18cfa1b1e40de725973184ceefb1165/390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java#L34)):*
> ```Java
>     log.info("Starting Service Account Token migration");
>     String accountId =
>         "yTbabkw4SdCPZdnEWynUNg"; // Account Id for the customer- "Kroger" for whom the migration is required
> ```

Instead of hardcoding accountId here can we pick this up via FF?

File: 390-db-migration/src/main/java/io/harness/migrations/MigrationList.java [hunk 1]
```diff
@@ -454,6 +455,7 @@ public static List<Pair<Integer, Class<? extends Migration>>> getMigrations() {
         .add(Pair.of(372, BaseMigration.class))
         .add(Pair.of(373, AddRingDetailsToDelegateRing.class))
         .add(Pair.of(374, AddIsProductLedToAccount.class))
+        .add(Pair.of(375, SettingAttributesServiceAccountTokenMigration.class))
```
Comment (meenakshi-harness) [inline 390-db-migration/src/main/java/io/harness/migrations/MigrationList.java:458]: We can have this as Background Migration.

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 5 file(s) to review. 3 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java` line 34 at r2](https://reviewable.io/reviews/harness/harness-core/37231#-NBRl17n1n_dzOcf8DWD:-NC-9lBC27I8cenmrCei:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/efbb95c7e18cfa1b1e40de725973184ceefb1165/390-db-migration/src/main/java/io/harness/migrations/all/SettingAttributesServiceAccountTokenMigration.java#L34)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Instead of hardcoding accountId here can we pick this up via FF?
</blockquote></details>

Done.
___
*[`390-db-migration/src/main/java/io/harness/migrations/MigrationList.java` line 458 at r2](https://reviewable.io/reviews/harness/harness-core/37231#-NBJ6IrvsgWFAMBdMbrO-r2-458:-NC-9m4F7Bu_joqnZpUL:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/efbb95c7e18cfa1b1e40de725973184ceefb1165/390-db-migration/src/main/java/io/harness/migrations/MigrationList.java#L458)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

We can have this as Background Migration. 
</blockquote></details>

Done.

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 5 file(s) to review. 2 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 1 change requests. 2. unresolved discussions.

**Review by gn-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: 1. unresolved discussions.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37231)* status: :shipit: complete! ✓

**Pull request merged**